### PR TITLE
Fix OOM in multi-dim hist depending on argument order

### LINKED ISSUE #3948: OOM when histogramming depending on argument order
Let's say there is a binned data array with two coordinates, one lives "on the binning" (`x`) and one lives "on the events" (`y`) and the intention of the user is to create a 2D histogram over y and x.

If the outer binning is large it is easy to run into a situation where performance drops drastically, or the process runs out of memory, for example:

```python
import scipp as sc

# 1 million bins
da = sc.data.binned_x(1000_000, 1000_000)

da.hist(x=200, y=200, dim=da.dims)
# This works.

# Different argument order
da.hist(y=200, x=200, dim=da.dims)
# This crashes the kernel (I assume it OOMs because it tries to create a large intermediate binning).
```

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #3948.

`hist` with more than one set of edges bins by all but the last edges and histograms the last one. Edges that replace an existing dim shrink that dim in the binning step, whereas a dim not touched by it survives at its original length. Deferring such edges to the final histogramming step therefore produced an intermediate with a bin count given by the product of the original and the new dim lengths — 1e6 x 200 bins for the example in the issue, which OOMs, while the same call with the edges given in the other order was fine.

Edges are now ordered such that those replacing an existing dim are applied first, and the output is transposed back to the requested dim order. On `sc.data.binned_x(1e5, 1e5).hist(y=200, x=200, dim=da.dims)` this goes from 4.7 s and 3964 MiB peak to 0.04 s and 19 MiB; the 1e6 example from the issue runs in 0.3 s and is bit-identical to the working argument order.

Reordering is not always safe, so it is skipped in two cases: re-binning a dim drops coords defined over that dim, and histogramming by an outer coord erases that coord's dims — the latter would silently change which dims the result has, not just their order. In both cases the given order is kept.

The reordering also exposed a latent bug in the re-binning fallback added in #3839, reachable on `main` without any reordering: restoring the input dim order after `combine_bins` raised `DimensionError` whenever dims were erased or added. Fixed here.

Side effect worth knowing: the output dim order now always follows the requested order. Previously `hist(y=5, x=7, z=3)` could come back as `('x', 'y', 'z')`, since `make_binned` canonicalizes dim order internally. The new behavior is what the docstring examples already imply.

Two classes of pre-existing problems turned up along the way and are marked as strict `xfail` here rather than fixed, so that a future fix does not go unnoticed:

- #3952: `hist` remains order-dependent when none of the edges replaces an existing dim, so no ordering helps.
- #3955: `bin` and `group` accept a bin-edge outer coord that the C++ implementation rejects. #3956 is stacked on this PR and fixes it, since doing so is a user-visible break that deserves its own change.

## Test plan

- New tests cover order-invariance of 2-D `hist` for both event-coord and outer-only-coord binning, dim order over all permutations of a 3-D `hist`, and a large input that would need hundreds of GByte with the old ordering.
- `hist`, `bin` and `group` were cross-checked against pre-PR behaviour by a differential run over 8 data layouts x all edge combinations, argument orders and `dim` values — 820 `hist` cases and 760 `bin`/`group` cases, comparing sizes, coords, masks, unit and a position-weighted checksum, with an independent reference that broadcasts the outer coord onto the events. Previously-crashing cases now work, with no regressions. Not committed, this was a one-off check.

### CODE REVIEW

**Comment by jokasimr** [comment]
Here are some test cases that are failing in this branch but run in the latest released version (unless the example says something else):

```python
import scipp as sc

da = sc.data.table_xyz(100).bin(x=2, y=3)
del da.bins.coords['x']
da.coords['x'] = sc.midpoints(da.coords['x'])

da.hist(z=2, x=2, dim=da.dims)
# Raises DimensionError.
```

```python
import scipp as sc

da = sc.data.binned_x(nevent=100, nbin=12)
da.coords['p'] = sc.midpoints(da.coords['x'])

da.hist(y=4, p=3, x=4, dim=da.dims)
# Raises KeyError for 'p'.
```

```python
import scipp as sc

da = sc.data.table_xyz(1200).bin(x=40, y=30)
da.coords['p'] = sc.arange('x', 40, unit='s')
da.coords['r'] = (
    sc.arange('x', 40, unit='K') + 40 * sc.arange('y', 30, unit='K')
)

da.hist(p=20, z=20, r=20, dim=da.dims)
# Large allocation: 40 * 30 * 20 * 20 = 480,000 intermediate bins.
# This issue also existed before this PR.
```

```python
import scipp as sc

da = sc.data.table_xyz(100).bin(x=2, y=3)
del da.bins.coords['y']

da.hist(z=2, y=2, dim=da.dims)
# Raises KeyError for 'y'.
# Before the PR this raised BinEdgeError. 
```

**Comment by SimonHeybrock** [comment]
Thanks, these were very useful — all four reproduce.

The first two were my regression: the reordering did not account for re-binning a dim dropping coords defined over that dim, nor for histogramming by an outer coord erasing that coord's dims (the latter would have silently changed which dims the result has, not just their order). Both now fall back to the given order.

Your case 1 and case 4 also turned out to be latent bugs in `make_binned` from #3839, reachable on `main` without any reordering — on the same inputs, `da.bin(x=2, dim=da.dims)` raises `DimensionError` and `da.bin(y=2, dim=da.dims)` raises `KeyError`. Fixed those rather than steering around them.

I replaced my ad-hoc checks with a differential comparison against pre-PR behaviour over 8 data layouts x all edge combinations, argument orders and `dim` values — 820 `hist` cases plus 760 `bin`/`group` cases, comparing sizes, coords, masks, unit and a position-weighted checksum. It caught 5 further regressions beyond the ones you found. Current state: no regressions, 42 previously-crashing cases work (checked against a reference that broadcasts the outer coord onto the events), and 28 confusing `KeyError`s are now proper `BinEdgeError`/`DimensionError`.

Your case 3 I left alone, since as you say it predates this PR and no ordering helps: none of `p`/`z`/`r` replaces an existing dim, so the input dims survive the binning step either way. I opened #3952 for it, together with two other pre-existing order-dependent cases the differential run turned up. The latter two are marked as strict xfail here so a future fix does not go unnoticed.

File: tests/binning_test.py [hunk 1]
```diff
@@ -1675,3 +1675,165 @@ def test_bin_rebin_existing_edge_cases() -> None:
...
+@pytest.mark.xfail(
+    reason="scipp/scipp#3952: hist by a 1-D and a 2-D outer coord depends on the "
+    "argument order",
+    strict=True,
+)
+def test_hist_by_1d_and_2d_outer_coords_is_order_independent() -> None:
+    da = sc.data.table_xyz(1000).bin(x=5, y=4)
+    da.coords['p'] = sc.arange('x', 5, unit='s')
+    da.coords['r'] = sc.arange('x', 5, unit='K') + 5 * sc.arange('y', 4, unit='K')
+    expected = da.hist(r=2, p=3, dim=())
```
Comment (jokasimr) [inline tests/binning_test.py:1863]: This test only tests that the result is order independent, but the results can still be incorrect.

This test fails
```python
def test_hist_by_1d_and_2d_outer_coords_preserves_existing_dims() -> None:
      da = sc.data.table_xyz(4).bin(x=2, y=2)
      da.coords['p'] = sc.arange('x', 2, unit='s')
      da.coords['r'] = (
          sc.arange('x', 2, unit='K')
          + 2 * sc.arange('y', 2, unit='K')
      )
      result = da.hist(r=2, p=2, dim=())
      assert result.sizes == {'x': 2, 'y': 2, 'r': 2, 'p': 2}
```

but I think it should succeed.

Comment (jokasimr) [reply tests/binning_test.py:1863]: Maybe we need to change/simplify the semantics of the `dims=` argument. It seems to me that we are running into a lot of edge cases (that probably are used very rarely) and it might be better to just explicitly disallow some of those cases.

As I understand it the `dims` argument was mainly introduced to avoid keeping (some or all) existing dimensions of the binning when histogramming or binning by new coordinates.
That avoided having to do a `.concat()` before the histogramming, and was a nice usability improvement.
But it also introduced the option of binning in coordinates with **the same dimensions** as the existing binning **while keeping the existing binning** (by passing `dims=()`).
That was new functionality that was not possible previously, because before `dim=` was introduced `.hist` always summed over the existing dimensions if the new binning coordinate shared dimensions with the existing binning. (As I understand it)

Seems to me that we could say `.hist` (and `.bin`) always sums over (or concatenates) existing binnings that share dimensions with the new binning coordinates, whatever is passed in `dims=`, and that would still keep the nice functionality of being able to concatenate some of the existing dimensions.
So basically, the semantics of `dim=` would then be "what **more** dimensions to sum/concat over, **beyond** the ones that the binning coordinates (the coordinates referenced by the arguments to `.hist` or `.bin`) refer to.

Comment (SimonHeybrock) [reply tests/binning_test.py:1863]: First, see also [ADR-0018](https://scipp.github.io/development/adr/0018-bin-hist-reduction-dims.html).

> Seems to me that we could say `.hist` (and `.bin`) always sums over (or concatenates) existing binnings that share dimensions with the new binning coordinates, whatever is passed in `dims=`

That was more or less how it worked before the ADR. I am not sure what the non-explicit definition (implicit by which dims the coord has) of which dims to sum/concat over would add. On the contrary, it seems it would make certain things impossible, e.g., if you have `dims=('pixel', wavelength')` with a 2D `Q` coord and want to turn that into `('pixel', 'Q')` &mdash; if I understand your definition correctly you would always sum over `pixel`.

Comment (jokasimr) [reply tests/binning_test.py:1863]: >  if I understand your definition correctly you would always sum over pixel

Yes that's right. Yeah I see, removing the  option to avoid that would be limiting and not what we want.

>  I am not sure what the non-explicit definition (implicit by which dims the coord has) of which dims to sum/concat over would add. 

It would reduce the large increase in number of cases to handle, which is also the main drawback listed in the ADR.

Comment (SimonHeybrock) [reply tests/binning_test.py:1863]: Regarding the failing test: This seems to be the same behavior as on `main`, but it is inconsistent with the ADR, I'll open a stacked PR. Should not block the OOM fix in this branch.

**Comment by SimonHeybrock** [comment]
@jokasimr Do you have more comments about this fix?

**Comment by jokasimr** [comment]
I haven't reviewed the details yet, but I'll do it today 👍

**Review by jokasimr — APPROVED** [review]
